test/e2e/cache_watch_test.go

```diff
@@ -186,123 +221,140 @@ var _ = Describe("Reconciles test:", func() {
+		}),
 	)
 })
 
-func setupNamespaces(testData *model.TestDataProvider, name string, size int) []string {
-	namespaces := make([]string, size)
+func setupNamespaces(ctx context.Context, testData *model.TestDataProvider, generateNames ...string) []*corev1.Namespace {
+	result := make([]*corev1.Namespace, 0, len(generateNames))
 	By("Setting up test namespaces (and wait for them to be created)", func() {
-		for i := 0; i < size; i++ {
-			namespaces[i] = utils.RandomName(fmt.Sprintf("%s-ns%d", name, i))
-			Expect(k8s.CreateNamespace(testData.Context, testData.K8SClient, namespaces[i])).To(Succeed())
+		for i := 0; i < len(generateNames); i++ {
+			namespace, err := k8s.CreateRandomNamespace(ctx, testData.K8SClient, generateNames[i])
+			Expect(err).To(BeNil())
+			result = append(result, namespace)
 		}
-		for _, ns := range namespaces {
+		for _, ns := range result {
 			Eventually(func(g Gomega) bool {
 				namespace := &corev1.Namespace{}
 				return g.Expect(
-					testData.K8SClient.Get(testData.Context, types.NamespacedName{Name: ns}, namespace),
+					testData.K8SClient.Get(ctx, types.NamespacedName{Name: ns.GetName()}, namespace),
 				).To(Succeed())
 			}).WithTimeout(time.Minute).Should(BeTrue())
 		}
 	})
-	return namespaces
+	return result
 }
 
-func setupSecrets(testData *model.TestDataProvider, namespaces []string) {
+func setupSecrets(ctx context.Context, testData *model.TestDataProvider, namespaces []*corev1.Namespace) {
 	By("Setting up test secrets, one on each namespace (and wait for them)", func() {
 		for _, ns := range namespaces {
-			k8s.CreateDefaultSecret(testData.Context, testData.K8SClient, config.DefaultOperatorGlobalKey, ns)
+			k8s.CreateDefaultSecret(ctx, testData.K8SClient, config.DefaultOperatorGlobalKey, ns.GetName())
 		}
 		for _, ns := range namespaces {
 			Eventually(func(g Gomega) bool {
 				secret := &corev1.Secret{}
 				return g.Expect(
-					testData.K8SClient.Get(testData.Context, types.NamespacedName{
-						Name: config.DefaultOperatorGlobalKey, Namespace: ns}, secret),
+					testData.K8SClient.Get(ctx, types.NamespacedName{
+						Name: config.DefaultOperatorGlobalKey, Namespace: ns.GetName()}, secret),
 				).To(Succeed())
 			}).WithTimeout(time.Minute).Should(BeTrue())
 		}
 	})
 }
 
-func setupManager(testData *model.TestDataProvider, namespaces []string, namespaceIndexesToWatch []int) *managerHandle {
-	mgrHandle := managerHandle{}
-	mgrCtx, cancelFn := context.WithCancel(context.Background())
-	mgrHandle.cancelFn = cancelFn
+type stopper func()
+
+func setupManager(ctx context.Context, namespaces []*corev1.Namespace, wantToWatch []string) (manager.Manager, stopper) {
+	var (
+		wg  sync.WaitGroup
+		mgr manager.Manager
+	)
+
+	wantToWatchSet := sets.NewString(wantToWatch...)
+	watchedNamespaces := make(map[string]bool)
+	for _, ns := range namespaces {
+		if wantToWatchSet.Has(ns.GetGenerateName()) {
+			watchedNamespaces[ns.GetName()] = true
+		}
+	}
```

Comment:
note: I am not convinced we need this here. I reimplemented it as it was there before but I think we can simply `managerConfig.WatchedNamespaces = wantToWatch` and call it a day. Feel free to remove this.